### PR TITLE
User Profile - Add new action origin and internal user

### PR DESCRIPTION
**Pull request by ywangd**
Profile documents are stored in a separate system index from the main
security index. Hence a more scoped origin and internal user is better
than the all powerful _xpack_security user. This PR adds
_security_profile user that has privileges only over the profile index
and updates all profile related actions to use it.

### CODE REVIEW

**Comment by ywangd** [comment]
Some more context: We discussed previously that internal users could be of its own subtype of `User`. I thought Java's sealed class and pattern matching for switch would be great for this refactor. However, pattern matching for switch is still a preview feature for Java 17 (and 18). Without it, the refactoring would be less interesting. Anyway, I think the refactoring is better handled separately regardless.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/user/SecurityProfileUser.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+    public static final String NAME = UsernamesField.SECURITY_PROFILE_NAME;
+    public static final SecurityProfileUser INSTANCE = new SecurityProfileUser();
+    private static final String ROLE_NAME = UsernamesField.SECURITY_PROFILE_ROLE;
+    public static final RoleDescriptor ROLE_DESCRIPTOR = new RoleDescriptor(
+        ROLE_NAME,
+        null,
+        new RoleDescriptor.IndicesPrivileges[] {
+            RoleDescriptor.IndicesPrivileges.builder()
+                .indices(".security-profile*")
```
Comment (albertzaharovits) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/user/SecurityProfileUser.java:27]: I think we should use the same index pattern from the `SecuritySystemIndices#getSecurityProfileIndexDescriptor()` .

**Review by albertzaharovits — APPROVED** [review]
LGTM

**Review by n1v0lg — APPROVED** [review]
LGTM

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/user/SecurityProfileUser.java (hunk 1 above)
Comment (ywangd) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/user/SecurityProfileUser.java:27]: I didn't want to use regex because I was paranoid about performance. But it's probably trivial (since XPackUser has them). So I updated to match the pattern used by the system index.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/user/SecurityProfileUser.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+ */
+package org.elasticsearch.xpack.core.security.user;
+
+import org.elasticsearch.xpack.core.security.authz.RoleDescriptor;
+import org.elasticsearch.xpack.core.security.support.MetadataUtils;
+
+import java.util.Map;
+
+/**
+ * internal user that manages xpack security. Has all cluster/indices permissions.
```
[the remark this answers is not in the document]
Comment (ywangd) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/user/SecurityProfileUser.java:15]: Good catch! thanks

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/profile/ProfileServiceTests.java [hunk 3]
```diff
@@ -306,6 +320,58 @@ public void testBuildSearchRequest() {
...
+        // Update
+        doAnswer(invocation -> {
+            assertThat(threadPool.getThreadContext().getTransient(ACTION_ORIGIN_TRANSIENT_NAME), equalTo(SECURITY_PROFILE_ORIGIN));
+            final ActionListener<?> listener = (ActionListener<?>) invocation.getArguments()[2];
+            listener.onFailure(expectedException);
+            return null;
+        }).when(client).execute(eq(UpdateAction.INSTANCE), any(UpdateRequest.class), anyActionListener());
+        final PlainActionFuture<UpdateResponse> future2 = new PlainActionFuture<>();
+        profileService.doUpdate(mock(UpdateRequest.class), future2);
+        final RuntimeException e2 = expectThrows(RuntimeException.class, future2::actionGet);
```
[the remark this answers is not in the document]
Comment (ywangd) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/profile/ProfileServiceTests.java:359]: What we actually want to test is the action origin is switched to `security_profile` for each profile related actions. This is done by the assertion in the `doAnswer`. What the lambda actually returns is irrelevant after the assertion is passed. Mocking a full successful response is possible but not useful here since they are no different than an exception with what the test tries to achieve. So I chose to keep it simple. But to address your point, I add a comment to the test method to explain what it does.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/user/UsernamesField.java [hunk 4]
```diff
@@ -16,6 +16,8 @@ public final class UsernamesField {
     public static final String SYSTEM_ROLE = "_system";
     public static final String XPACK_SECURITY_NAME = "_xpack_security";
     public static final String XPACK_SECURITY_ROLE = "_xpack_security";
+    public static final String SECURITY_PROFILE_NAME = "_security_profile";
```
[the remark this answers is not in the document]
Comment (ywangd) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/user/UsernamesField.java:19]: This is a great point! I thought about that one can keep using any existing `_security_profile` user without it clashing with the new internal user because role resolving for internal users has its own distinct path. But you are right that adding a new internal user has more impact on the user-facing side:
1. It will no longer be possible to add a new native realm user named `_security_profile`
2. If there is an existing `_security_profile` native realm user, it will not be possible to update it, including changing password or deleting it. But otherwise it will work normally.

@tvernum and I discussed and we don't think it should be considered breaking because we never considered adding new internal users breaking in the past. We don't want to change the existing stance because it means we will never be able to add another internal users.

That said, there is a discussion about how much validation we should have for username and more broadly the future of internal users. Today, you can create a `_xpack_security` using the file realm. This usename can also comes with external realms (LDAP etc) where we have no control over usernames. So why should native realm specifically bother with it especially since internally an "externally" defined `_xpack_security` will not be confused with the true internal user? 

In summary, the plan I will go with is:
1. Keep this PR as is.
2. Start a separate discussion about options to make adding internal users less impactful (aiming for 8.3 as a follow-up for this PR) and future direction of internal users.

[the remark this answers is not in the document]
Comment (n1v0lg) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/user/UsernamesField.java:19]: Agreed with the next steps! Adding my thoughts to the slack thread https://elastic.slack.com/archives/CA6N3HDF1/p1650527946891839
